### PR TITLE
Type-safe Security Configuration

### PR DESCRIPTION
**Pull request by felixschlegel**
> **Note**: this PR sits on top of #87.

> Adds some of the missing properties from #46.

### Motivation:

We were lacking some of the security protocol configuration options and
want to provide them in a type-safe manner.

### Modifications:

* add Kerberos support
* add OAuthBearer support
* Created a new file `KafkaConfiguration+Security.swift`
    * add `public struct` `SSLConfiguration`
    * add `public struct` `SASLMechanism`
    * add `public struct` `KerberosConfiguration`
    * add `public struct` `OAuthBearerMethod`
    * add `public struct` `SecurityProtocol`
* integrate `KafkaConfiguration.SSLOptions` in new types
* integrate `KafkaConfiguration.SASLOptions` in new types

### Example:

```swift
var config = KafkaConsumerConfiguration(
    consumptionStrategy: .partition(.unassigned, topic: "some topic")
)
config.securityProtocol = .saslSSL(
    saslMechanism: .gssapi(kerberosConfiguration: .init(keytab: "/path/to/keytab")),
    sslConfiguaration: .keyPair(
        privateKey: .init(
            location: .pem("jfklaffalk"),
            password: "iLoveApple"
        ),
        publicKeyCertificate: .file(location: "/path/to/file"),
        caCertificate: .pem("jfkaljfl"),
        crlLocation: nil
    )
)
```

### CODE REVIEW

File: Sources/SwiftKafka/Configuration/KafkaConfiguration+Security.swift [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+        /// - Parameters:
+        ///
+        ///     - privateKey: The client's private key (PEM) used for authentication.
+        ///     - publicKeyCertificate: The client's public key (PEM) used for authentication.
+        ///     - caCertificate: File or directory path to CA certificate(s) for verifying the broker's key.
+        ///     - crlocation: Path to CRL for verifying broker's certificate validity.
+        public static func keyPair(
+            privateKey: PrivateKey,
+            publicKeyCertificate: Key,
+            caCertificate: Key,
```
Comment (dnadoba) [inline Sources/SwiftKafka/Configuration/KafkaConfiguration+Security.swift:97]: I think we should provide an easy way to use the "default" i.e. "probe" that searches for locally trusted CAs.

File: Sources/SwiftKafka/Configuration/KafkaConfiguration+Security.swift [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+extension KafkaConfiguration {
+    // MARK: - SSLConfiguration
+
+    /// Use to configure an SSL connection.
+    public struct SSLConfiguration: Sendable, Hashable {
+        /// A key either located in a file or directly as a key String (PEM format).
+        public struct Key: Sendable, Hashable {
```
Comment (dnadoba) [inline Sources/SwiftKafka/Configuration/KafkaConfiguration+Security.swift:21]: Probably the right term here would be `Chain` as it not only contains the public key but at least one certificate and possibly multiple certificates.

File: Sources/SwiftKafka/Configuration/KafkaConfiguration+Security.swift (hunk 1 above)
Comment (dnadoba) [inline Sources/SwiftKafka/Configuration/KafkaConfiguration+Security.swift:97]: we might want to use different types here so we can provide a reasonable default for `caCertificate`.

File: Sources/SwiftKafka/Configuration/KafkaConfiguration+Security.swift (hunk 2 above)
Comment (dnadoba) [reply Sources/SwiftKafka/Configuration/KafkaConfiguration+Security.swift:21]: Hm... maybe even two types.
- One of them is kind of a `LeafAndIntermediates` type as the first certificate in the chain will be used as the identity of the client and the remaining, in no particular order, will be used as potential intermediate certificates and send to the broker to verify the client.
- The other type could be `RootCertificates` as it is an unordered list of trusted CAs that is used by the client to verifier the identity of the broker.

**Review by FranzBusch — APPROVED** [review]

File: Sources/SwiftKafka/Configuration/KafkaConfiguration+Security.swift [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+            return resultDict
+        }
+    }
+
+    // MARK: - SASLMechanism
+
+    /// Available SASL mechanisms that can be used for authentication.
+    public struct SASLMechanism: Sendable, Hashable {
+        /// Used to configure Kerberos.
+        public struct KerberosConfiguration: Sendable, Hashable {
```
Comment (dnadoba) [inline Sources/SwiftKafka/Configuration/KafkaConfiguration+Security.swift:242]: I think there is a public init missing

File: Sources/SwiftKafka/Configuration/KafkaConfiguration+Security.swift [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+
+        private let _internal: _SecurityProtocol
+
+        /// Send messages as plaintext (no security protocol used).
+        public static let plaintext = SecurityProtocol(
+            _internal: .plaintext
+        )
+
+        /// Use the Secure Sockets Layer (SSL) protocol.
+        public static func ssl(configuration: SSLConfiguration) -> SecurityProtocol {
```
Comment (dnadoba) [inline Sources/SwiftKafka/Configuration/KafkaConfiguration+Security.swift:456]: @lukasa should this be called TLS instead of SSL?

**Review by dnadoba — COMMENTED** [review]
Do we allow users to modify the config dictionary? e.g. to allow users using config options the swift wrappers do not yet support?

**Review by dnadoba — APPROVED** [review]
Awesome! Looks good to me

**Pull request merged**